### PR TITLE
Deepensemble Pipeline

### PR DESCRIPTION
**Pull request by prajurock**
# Description

Evaluation for the Deep Ensemble approach. 

User story #790 (https://dev.azure.com/cgmorg/ChildGrowthMonitor/_workitems/edit/790

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Submitted the evaluation task from my local machine to AzureML.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests
- [ ] New and existing unit tests pass locally with my changes
- [ ] My Code is review by 1 People

### CODE REVIEW

**Review by markus-hinsche — COMMENTED** [review]
Please look at my review here: https://github.com/Welthungerhilfe/cgm-ml/pull/269

**Comment by AI-Guru** [comment]
Great work!

An observation: There are not two places where models can get downloaded. evaluate_main.py and evaluate.py.

I believe it would be best practice to only download in evaluate.py. Pushing snapshots to AzureML is a bottleneck and this would get rid of that. I think this has a high priority.

Another observation: When branching between a single model and multiple models, you check for RUN_ID == "all" in multiple places.

I guess it would be better to check if RUN_ID is "all" or a list of RUN_IDs for the ensemble mode. This would allow us to select specific models for ensembling. Rather low prior but definitely a nice to have.

Looks really great!

**Comment by markus-hinsche** [comment]
> An observation: There are not two places where models can get downloaded. evaluate_main.py and evaluate.py.

This would be nice I agree. We could also do this in a follow-up PR after this one I think

**Comment by AI-Guru** [comment]
> > An observation: There are not two places where models can get downloaded. evaluate_main.py and evaluate.py.
> 
> This would be nice I agree. We could also do this in a follow-up PR after this one I think

Fine with me! Should be a quick done.

File: src/common/evaluation/QA/eval_depthmap_models/src/qa_config_height_deep_ensemble.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import os
+
+from bunch import Bunch
+
+CONFIG_NAME = os.path.splitext(os.path.basename(__file__))[0]
+
+# Details of model used for evaluation
+MODEL_CONFIG = Bunch(dict(
+    EXPERIMENT_NAME='q1-ensemble-warmup',
+    RUN_IDs=['q1-ensemble-warmup_1610544610_eb44bfe2', 'q1-ensemble-warmup_1610547587_7ca932c3',
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/qa_config_height_deep_ensemble.py:10]:
```suggestion
    RUN_IDS=['q1-ensemble-warmup_1610544610_eb44bfe2', 'q1-ensemble-warmup_1610547587_7ca932c3',
```

File: src/common/evaluation/QA/eval_depthmap_models/eval_main.py [hunk 2]
```diff
@@ -74,11 +75,12 @@ def copy_dir(src: Path, tgt: Path, glob_pattern: str, should_touch_init: bool =
     MODEL_BASE_DIR.mkdir(parents=True, exist_ok=True)
 
     # Copy model to temp folder
-    download_model(workspace,
-                   experiment_name=MODEL_CONFIG.EXPERIMENT_NAME,
-                   run_id=MODEL_CONFIG.RUN_ID,
-                   input_location=os.path.join(MODEL_CONFIG.INPUT_LOCATION, MODEL_CONFIG.NAME),
-                   output_location=MODEL_BASE_DIR)
+    if MODEL_CONFIG.RUN_ID != 'all':
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/eval_main.py:78]: 'all' is not used anymore in the config

File: src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py [hunk 3]
```diff
@@ -174,6 +177,17 @@ def get_prediction(model_path: str, dataset_evaluation: tf.data.Dataset) -> np.a
     return prediction_list
 
 
+def get_predictions(model_paths: list, dataset_evaluation: tf.data.Dataset) -> list:
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py:180]: The function name can be more distinctive to `get_prediction()`

```suggestion
def get_predictions_from_multiple_models(model_paths: list, dataset_evaluation: tf.data.Dataset) -> list:
```

File: src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py [hunk 4]
```diff
@@ -87,7 +89,8 @@ def copy_dir(src: Path, tgt: Path, glob_pattern: str, should_touch_init: bool =
 FILTER_CONFIG = qa_config.FILTER_CONFIG if getattr(qa_config, 'FILTER_CONFIG', False) else None
 
 
-RUN_ID = MODEL_CONFIG.RUN_ID
+RUN_ID = MODEL_CONFIG.RUN_ID if getattr(MODEL_CONFIG, 'RUN_ID', False) else None
+RUN_IDs = MODEL_CONFIG.RUN_IDs if getattr(MODEL_CONFIG, 'RUN_IDs', False) else None
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/evaluate.py:93]: Maybe "xor" could help to make sure exactly one is set

```suggestion
RUN_ID = MODEL_CONFIG.RUN_ID if getattr(MODEL_CONFIG, 'RUN_ID', False) else None
RUN_IDs = MODEL_CONFIG.RUN_IDs if getattr(MODEL_CONFIG, 'RUN_IDs', False) else None
bool(RUN_ID) != bool(RUN_IDS)
```

File: src/common/evaluation/QA/eval_depthmap_models/src/qa_config_height_deep_ensemble.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    EXPERIMENT_NAME='q1-ensemble-warmup',
+    RUN_IDS=['q1-ensemble-warmup_1610544610_eb44bfe2', 'q1-ensemble-warmup_1610547587_7ca932c3',
+             'q1-ensemble-warmup_1610547669_5b789bd1', 'q1-ensemble-warmup_1610547705_f2141d0f',
+             'q1-ensemble-warmup_1610547744_d2b42ce5', 'q1-ensemble-warmup_1610547780_2f000a25',
+             'q1-ensemble-warmup_1610547816_c3f815df', 'q1-ensemble-warmup_1610547892_8ee6ff49',
+             'q1-ensemble-warmup_1610547928_b9519b6a', 'q1-ensemble-warmup_1610547986_ad0186b8',
+             'q1-ensemble-warmup_1610548023_99ac6060', 'q1-ensemble-warmup_1610548064_afefd4e4',
+             'q1-ensemble-warmup_1610548106_69993d24', 'q1-ensemble-warmup_1610548137_a8c52d63',
+             'q1-ensemble-warmup_1610548168_914ce1f6', 'q1-ensemble-warmup_1610548209_9692a253'],
+
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/qa_config_height_deep_ensemble.py:18]:
```suggestion
```

File: src/common/evaluation/uncertainty_utils.py [hunk 6]
```diff
@@ -59,3 +59,32 @@ def predict_uncertainty(X: np.array, model: tf.keras.Model, num_dropout_predicti
     predictions = model(one_batch, training=True)
     std = tf.math.reduce_std(predictions)
     return std
+
+
+def get_prediction_uncertainty_deep(model_paths: list, dataset_evaluation: tf.data.Dataset) -> np.array:
```
Comment (markus-hinsche) [inline src/common/evaluation/uncertainty_utils.py:64]:
```suggestion
def get_prediction_uncertainty_deepensemble(model_paths: list, dataset_evaluation: tf.data.Dataset) -> np.array:
```

File: src/common/evaluation/uncertainty_utils.py [hunk 7]
```diff
@@ -59,3 +59,32 @@ def predict_uncertainty(X: np.array, model: tf.keras.Model, num_dropout_predicti
...
+    std_list = []
+    for model_path in model_paths:
+        print(f"loading model from {model_path}")
+        model = load_model(model_path, compile=False)
+        std_list += [[model.predict(X)[0] for X, y in dataset.as_numpy_iterator()]]
+    std_list = np.array(std_list)
+    std_list = np.std(std_list, axis=0)
+    std_list = std_list.reshape((-1))
+    end = time.time()
+    print(f"Total time for uncertainty prediction experiment: {end - start:.3} sec")
```
Comment (markus-hinsche) [inline src/common/evaluation/uncertainty_utils.py:88]: Please replace all printing in this PR by logging

**Pull request merged**